From b96e097ee9b56d9069e1b3b1771878f8ee01511c Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Sat, 22 May 2021 09:54:30 +0300
Subject: [PATCH] [SW-43026] module_fwupd: disable pcie on fit cmd

PCIe disabling mechanism is using a fit as a command for
device to close any check on fw_upd module and to stop
any further process of fw_upd module if given through pcie.
This is used as protection and security mechanism in order to
guarantee that fw_upd is disabled to pcie after the command
is received and will remain close till power cycle of device.
This fit is received using ELBI interface.
A next patch will save the information in boot_info and
also check boot_info on starting preboot, in order to
keep the closure between sw reset and PRSTN reset.
A later patch shall also use the state's flag in order
to disable access fw_udp from pcie.

Change-Id: I70fd476510ac018af6d290c0cfa4084621350859
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/common/cmd_module.h b/common/cmd_module.h
index 7a743b3..9faedbf 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -85,6 +85,7 @@
 void module_fwupd_run_fast(void);
 void module_fwupd_run(void);
 void module_fwupd_run_noret(void);
+void module_fwupd_pcie_disable(bool is_pcie_disabled);
 
 /* I2C FWUPD */
 void module_i2c_fwupd_setup(void);
diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index 177c170..c361fe2 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -42,6 +42,7 @@
 	void (*sts_writel)(u32 val);
 	u32 (*cmd_readl)(void);
 	bool use_security;
+	bool is_disabled;
 };
 
 static struct fwupd_state state;
@@ -358,6 +359,11 @@
 	state->sending_error_flag = false;
 }
 
+void module_fwupd_pcie_disable(bool is_pcie_disabled)
+{
+	state.is_disabled = is_pcie_disabled;
+}
+
 static enum hl_upd_sts module_fwupd_map(struct fwupd_state *state,
 					unsigned int size, enum hl_upd_cmd cmd)
 {
@@ -527,6 +533,9 @@
 		return module_fwupd_load_type_nic_en(state->map_offset);
 	case FIT_LOAD_TYPE_RMA_ERASE:
 		return rma_erase();
+	case FIT_LOAD_TYPE_FW_UPD_CLOSURE:
+		module_fwupd_pcie_disable(true);
+		break;
 	case FIT_LOAD_TYPE_RUN:
 		return module_fwupd_load_type_run(state, state->map_offset);
 	default:
